### PR TITLE
Improve handling of slash commands

### PR DESCRIPTION
**Pull request by davidvorona**

### CODE REVIEW

File: src/commands.ts [hunk 1]
```diff
@@ -39,33 +66,150 @@ const commands: Record<string, Command> = {
...
         },
+        build() {
+            const builder = new SlashCommandBuilder()
+                .setName('track')
+                .addStringOption(option =>
+                    option
+                        .setName('username')
+                        .setDescription('Username')
+                        .setRequired(true));
+            return buildSlashCommand(builder, this);
```
Comment (evanw555) [inline src/commands.ts:77]: There seems to be a lot of redundancy in these `build()` commands. Is there any way we can do this all in the `buildCommands` loop? Also, we can do `setName` programmatically using inferred info from the command name (ensures consistency), and `setDescription` using the `text` property. The only thing we'd need to define per-command is the list of args (i.e. "options")

Comment (evanw555) [reply src/commands.ts:77]: Also, seems weird to me that we're calling `buildSlashCommand` in the command itself. IMO we should have this be done in `CommandHandler.buildCommands`

File: src/bot.ts [hunk 2]
```diff
@@ -204,6 +204,13 @@ client.on('ready', async () => {
             await timeoutManager.registerTimeout(TimeoutType.WeeklyXpUpdate, getNextFridayEvening(), { pastStrategy: PastTimeoutStrategy.Invoke });
         }
 
+        // Register global slash commands on startup
+        if (client.application) {
```
Comment (evanw555) [inline src/bot.ts:208]: This can run without me having to define `guildId` and `clientId` in the auth file, right?

**Comment by evanw555** [comment]
Technically it's not really necessary because this isn't a library, but I bump the package.json version in each commit

File: src/bot.ts (hunk 2 above)
Comment (davidvorona) [reply src/bot.ts:208]: Correct, it doesn't need the guildId in the first place and the clientId is known.

File: src/command-handler.ts [hunk 3]
```diff
@@ -39,124 +35,97 @@ class CommandHandler {
...
+            }
+            if (command.privileged) {
                 CommandHandler.assertIsAdmin(interaction);
-                const guildId = CommandHandler.getInteractionGuildId(interaction);
-                await updateTrackingChannel(guildId, interaction.channelId);
-                const textChannel = interaction.channel as TextBasedChannel;
-                state.setTrackingChannel(guildId, textChannel);
-                await interaction.reply('Player experience updates will now be sent to this channel');
             }
+            command.execute && await command.execute(interaction);
```
Comment (evanw555) [inline src/command-handler.ts:122]: Can you make this a proper if-statement, and reply with a warning the else case?

File: src/command-reader.ts [hunk 4]
```diff
@@ -20,14 +20,14 @@ class CommandReader {
...
-            const commandInfo: Command = commands[command];
+            const commandInfo: Command = commands[command as CommandName];
             if (commandInfo.privileged && !state.isAdmin(msg.author.id)) {
                 msg.channel.send('You can\'t do that');
             } else if (commandInfo.failIfDisabled && state.isDisabled()) {
                 msg.channel.send('I can\'t do that while I\'m disabled');
             } else {
                 try {
-                    commandInfo.fn(msg, rawArgs, ...args);
+                    commandInfo.fn && commandInfo.fn(msg, rawArgs, ...args);
```
Comment (evanw555) [inline src/command-reader.ts:30]: Similar to above comment, can you make this a proper if-statement and reply with a warning in the else case?

**Review by evanw555 — APPROVED** [review]

**Pull request merged**